### PR TITLE
Add rudimentary streams

### PR DESCRIPTION
**Pull request by kriskowal**
This change paves the way for a Q-Connection port.

- Adds defer and delay to Promise, for convenience.
- Adds AsyncQueue, based on Mark's promise queue implementation.
- Adds Stream and Pipe {in, out} implementations that haul iterations through an entangeld pair of promise queues. This enables back-pressure, concurrency limits, and conclusion.
- Adds an adapter that builds {in, out} streams out of web platform and Node.js stream-like objects: web workers, web message ports, Chrome extension ports, and Node.js streams,

### CODE REVIEW

File: src/async-queue.js
Comment (erights) [inline src/async-queue.js:14]: Is ends.reject used anywhere?

File: src/miniq.js
Comment (erights) [inline src/miniq.js]: I like that you avoided assigning to fields of the deferred.

File: src/miniq.js
Comment (erights) [inline src/miniq.js]: Both .delay and .defer should be initialized by defineProperty rather than assignment.

File: src/async-stream.js
Comment (erights) [inline src/async-stream.js]: This is a good example of why it is important to initialize with defineProperty rather than assignment. This assignment overrides Function.prototype. If Function.prototype is a non-writable data property, then this override-by-assignment will fail with a thrown error.

File: src/miniq.js
Comment (erights) [inline src/miniq.js]: Noting that setTimeout is not part of the EcmaScript standard. Nor is it part of the proposed Frozen Realm primordials. Both it and your Promise.delay enables one to measure duration. Which is a problem we've already got. We should at least note these issues in a comment here.

File: src/async-queue.js
Comment (kriskowal) [reply src/async-queue.js:14]: In Stream.prototype.forEach. As elsewhere, it’s convenient but unnecessary.

File: src/async-stream.js
Comment (kriskowal) [reply src/async-stream.js]: What rule do you use to govern whether to write properties directly? In the revisions I’ve just posted, I assumed the rule, “Use define properties to augment constructors, functions, and prototypes.”, and generally avoided mutation.

File: src/miniq.js
Comment (kriskowal) [reply src/miniq.js]: I can remove them and think of more deterministic illustration. I’m using delay only for a casual illustration of the concurrency cap supported by streams, which causes the logs to read (30, 20, 10, 60, 50, 40).

File: src/miniq.js
Comment (kriskowal) [reply src/miniq.js]: When in Rome.

**Review by kriskowal — COMMENTED** [review]
I’ve pushed changes to address feedback.

File: src/async-queue.js
Comment (erights) [reply src/async-queue.js:14]: I don't get it. 
   * I don't see how the ends object could escape AsyncQueue. It seems safely encapsulated.
   * I looked, and forEach uses the .reject of a deferred, which makes sense. AFAICT, it does not use the .reject of an ends object.

**Comment by erights** [comment]
> What rule do you use to govern whether to write properties directly? In the revisions I’ve just posted, I assumed the rule, “Use define properties to augment constructors, functions, and prototypes.”, and generally avoided mutation.

Yeah, that's about right. Also, static "declarative" structure --- such as a configuration of objects to be initialized before "real code" starts running" --- should only be initialized with defineProperty. 

For dynamic computation, sometimes assignment is right and sometimes defineProperty is right. The litmus test: If the property might be an accessor property, or if it might be absent as own but inherit an accessor, do you want the "write" to invoke the setter if there is one, or do you want to override the accessor-nature of the property?

From outside an object, assignment is a "normal" way to interact with the object's API, where the provider cares if this invokes a setter or not, but the invoker leaves this interpretation of the abstract request up to the provider. By contract, "defineProperty" is meta --- it is about manipulating an object's public API rather than using it. It should typically only be done by the provider of the abstraction represented by the object, but not by that abstraction's clients.

**Review by erights — APPROVED** [review]
Since the nanoq base this is building on is badly broken, I'm approving this change as a worthy experiment that does not need to be fully understood before approval. We'll need to revisit all this again anyway (especially nanoq) before this is production ready.

Sorry that nanoq is in such bad shape. But at least it got us started.

**Pull request merged**
